relay-server/src/utils/envelope_context.rs

```diff
@@ -66,10 +60,44 @@ impl EnvelopeContext {
         self
     }
 
-    /// Records outcomes for all items stored in this context.
+    /// Records an outcome scoped to this envelope's context.
+    ///
+    /// This envelope context should be updated using [`update`](Self::update) soon after this
+    /// operation to ensure that subsequent outcomes are consistent.
+    pub fn track_outcome(&self, outcome: Outcome, category: DataCategory, quantity: usize) {
```

Comment:
I wonder if we could get rid of the `pub` here? As far as I can see `process_profiles` is the only caller, I guess we need it there because we emit outcomes for individual profiles?

Reply:
We will actually have to call this in more places, at least:
 - For sessions. Right now, we're dropping sessions without logging outcomes.
 - For the rate limiter once `enforcement.track_outcomes` is migrated to the context.

This is definitely a somewhat unsafe API in the interim.